File: android/app/src/test/java/com/mdreader/util/DateBucketsTest.kt
package com.mdreader.util

import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.Calendar

class DateBucketsTest {

    private val dayMs = 24L * 60 * 60 * 1000

    /** Today at the given hour, in the machine's default timezone. */
    private fun todayAt(hour: Int, minute: Int = 0): Long {
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, hour)
        cal.set(Calendar.MINUTE, minute)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    @Test
    fun sameDayIsToday() {
        val now = todayAt(13)
        assertEquals(DayBucket.TODAY, DateBuckets.bucket(now, now))
        assertEquals(DayBucket.TODAY, DateBuckets.bucket(todayAt(0, 5), now))
    }

    @Test
    fun previousDayIsYesterday() {
        val now = todayAt(13)
        assertEquals(DayBucket.YESTERDAY, DateBuckets.bucket(now - dayMs, now))
    }

    @Test
    fun threeDaysAgoIsEarlier() {
        val now = todayAt(13)
        assertEquals(DayBucket.EARLIER, DateBuckets.bucket(now - 3 * dayMs, now))
    }

    @Test
    fun formatProducesNonEmptyString() {
        val formatted = DateBuckets.format(todayAt(9, 30))
        assert(formatted.isNotBlank())
        assert(formatted.contains("/"))
    }
}

File: macos/Tests/DateBucketsTest.swift
import XCTest
@testable import MDreader

final class DateBucketsTest: XCTestCase {
    private let dayInterval: TimeInterval = 24 * 60 * 60

    private func date(hour: Int, minute: Int = 0, calendar: Calendar = .current) -> Date {
        var components = calendar.dateComponents([.year, .month, .day], from: Date())
        components.hour = hour
        components.minute = minute
        components.second = 0
        return calendar.date(from: components)!
    }

    func testSameDayIsToday() {
        let now = date(hour: 13)
        XCTAssertEqual(DateBuckets.bucket(now, now: now), .today)
        XCTAssertEqual(DateBuckets.bucket(date(hour: 0, minute: 5), now: now), .today)
    }

    func testPreviousDayIsYesterday() {
        let now = date(hour: 13)
        XCTAssertEqual(DateBuckets.bucket(now.addingTimeInterval(-dayInterval), now: now), .yesterday)
    }

    func testThreeDaysAgoIsEarlier() {
        let now = date(hour: 13)
        XCTAssertEqual(DateBuckets.bucket(now.addingTimeInterval(-3 * dayInterval), now: now), .earlier)
    }

    func testFormatProducesNonEmptyStringWithSlash() {
        let formatted = DateBuckets.format(date(hour: 9, minute: 30))
        XCTAssertFalse(formatted.isEmpty)
        XCTAssertTrue(formatted.contains("/"))
    }
}
